shell/platform/linux/fl_standard_message_codec_test.cc

```diff
@@ -1077,6 +1090,128 @@ TEST(FlStandardMessageCodecTest, DecodeUnknownType) {
+               fl_test_standard_message_codec_parent_class)
+        ->write_value(codec, buffer, value, error);
+  }
+}
+
+static FlValue* read_custom_value(FlStandardMessageCodec* codec,
+                                  GBytes* buffer,
+                                  size_t* offset,
+                                  GError** error) {
+  uint32_t length;
+  if (!fl_standard_message_codec_read_size(codec, buffer, offset, &length,
+                                           error)) {
+    return nullptr;
+  }
+  if (*offset + length > g_bytes_get_size(buffer)) {
+    g_set_error(error, FL_MESSAGE_CODEC_ERROR,
+                FL_MESSAGE_CODEC_ERROR_OUT_OF_DATA, "Unexpected end of data");
+    return nullptr;
+  }
+  FlValue* value = fl_value_new_custom(
+      128,
+      g_strndup(static_cast<const gchar*>(g_bytes_get_data(buffer, nullptr)) +
+                    *offset,
+                length),
+      g_free);
+  *offset += length;
+
+  return value;
+}
+
+static FlValue* fl_test_standard_message_codec_read_value_of_type(
+    FlStandardMessageCodec* codec,
+    GBytes* buffer,
+    size_t* offset,
+    int type,
+    GError** error) {
+  if (type == 128) {
+    return read_custom_value(codec, buffer, offset, error);
+  } else {
+    return FL_STANDARD_MESSAGE_CODEC_CLASS(
+               fl_test_standard_message_codec_parent_class)
+        ->read_value_of_type(codec, buffer, offset, type, error);
+  }
+}
+
+static void fl_test_standard_message_codec_class_init(
+    FlTestStandardMessageCodecClass* klass) {
+  FL_STANDARD_MESSAGE_CODEC_CLASS(klass)->write_value =
+      fl_test_standard_message_codec_write_value;
+  FL_STANDARD_MESSAGE_CODEC_CLASS(klass)->read_value_of_type =
+      fl_test_standard_message_codec_read_value_of_type;
+}
+
+static void fl_test_standard_message_codec_init(
+    FlTestStandardMessageCodec* self) {
+  // The following line suppresses a warning for unused function
+  FL_IS_TEST_STANDARD_MESSAGE_CODEC(self);
+}
+
+static FlTestStandardMessageCodec* fl_test_standard_message_codec_new() {
+  return FL_TEST_STANDARD_MESSAGE_CODEC(
+      g_object_new(fl_test_standard_message_codec_get_type(), nullptr));
+}
+
+TEST(FlStandardMessageCodecTest, DecodeCustomType) {
+  g_autoptr(FlTestStandardMessageCodec) codec =
+      fl_test_standard_message_codec_new();
+  g_autoptr(FlValue) value =
+      decode_message_with_codec("800568656c6c6f", FL_MESSAGE_CODEC(codec));
+  ASSERT_EQ(fl_value_get_type(value), FL_VALUE_TYPE_CUSTOM);
+  EXPECT_STREQ(static_cast<const gchar*>(fl_value_get_custom_value(value)),
+               "hello");
+}
+
+TEST(FlStandardMessageCodecTest, EncodeCustomType) {
+  g_autoptr(FlValue) value = fl_value_new_custom(128, "hello", nullptr);
+  g_autoptr(FlTestStandardMessageCodec) codec =
+      fl_test_standard_message_codec_new();
+  g_autofree gchar* hex_string =
+      encode_message_with_codec(value, FL_MESSAGE_CODEC(codec));
+  EXPECT_STREQ(hex_string, "800568656c6c6f");
+}
+
```

Comment:
Could we have a unit test with two custom types so I can get a clearer sense of how this works (and so that we can make sure that it works; right now the custom type's type happens to be the same as the magic custom type constant, which makes me nervous that we wouldn't catch a bug for anything using a non-128 value).

Reply:
Added en/decode tests that encode two customs types in a list.